Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/DataSynchronizer.swift

```diff
@@ -0,0 +1,557 @@
+            join.signal()
+        }
+        join.wait()
+        self.syncWorkItem = nil
+    }
+
+    /**
+     Requests that the given document _ids be synchronized.
+     - parameter ids: the document _ids to synchronize.
+     - parameter namespace: the namespace these documents belong to
+     */
+    func sync(ids: [BSONValue], in namespace: MongoNamespace) {
+        ids.forEach { id in
+            guard var nsConfig = syncConfig[namespace] else {
+                return
+            }
+
+            nsConfig.sync(id: id)
+        }
+
+        self.triggerListening(to: namespace)
+    }
+
+    /**
+     Stops synchronizing the given document _ids. Any uncommitted writes will be lost.
+
+     - parameter ids: the _ids of the documents to desynchronize.
+     */
+    func desync(ids: [BSONValue], in namespace: MongoNamespace) {
+        ids.forEach { id in
+            guard var namespace = syncConfig[namespace] else {
+                return
+            }
+
+            namespace[id] = nil
+        }
+
+        self.triggerListening(to: namespace)
+    }
+
+    /**
+     Returns the set of synchronized document ids in a namespace.
+
+     - returns: the set of synchronized document ids in a namespace.
+     */
+    func syncedIds(in namespace: MongoNamespace) -> Set<HashableBSONValue> {
+        guard let nsConfig = syncConfig[namespace] else {
+            return Set()
+        }
+
+        let v = nsConfig.map { $0.documentId }
+        print(v)
+        return Set(nsConfig.map({HashableBSONValue($0.documentId)}))
+    }
+
+    /**
+     Return the set of synchronized document _ids in a namespace
+     that have been paused due to an irrecoverable error.
+
+     - returns: the set of paused document _ids in a namespace
+     */
+    func pausedIds(in namespace: MongoNamespace) -> Set<HashableBSONValue> {
+        guard let nsConfig = syncConfig[namespace] else {
+            return Set()
+        }
+
+        return Set(nsConfig.compactMap { docConfig in
+            guard docConfig.isPaused else {
+                return nil
+            }
+            return HashableBSONValue(docConfig.documentId)
+        })
+    }
+
+    /**
+     A document that is paused no longer has remote updates applied to it.
+     Any local updates to this document cause it to be resumed. An example of pausing a document
+     is when a conflict is being resolved for that document and the handler throws an exception.
+
+     - parameter documentId: the id of the document to resume syncing
+     - returns: true if successfully resumed, false if the document
+     could not be found or there was an error resuming
+     */
+    func resumeSync(for documentId: BSONValue,
+                    in namespace: MongoNamespace) -> Bool {
+        guard var nsConfig = syncConfig[namespace],
+            var docConfig = nsConfig[documentId] else {
+                return false
+        }
+
+        docConfig.isPaused = false
```

Comment:
[minor discussion] So I think now that we have this skeleton a bit more fleshed out, it's clearer what happens to errors further down the stack. I know I suggested it but I'm not convinced yet that the `FatalErrorListener` is the right pattern for us everywhere. We don't have to resolve this on this particular PR (probably just when we merge `sync` into `master`), but wanted to bring it up so we can start think about it.

Setting docConfig to be paused can fail if the underlying config document can't be updated. In Android, this error is not actually squashed, `docConfig.setPaused` would simply throw, and the user would just get an exception when they call `resumeSyncForDocument` on the `CoreSync`. I think that's reasonable, especially considering that the error results directly from a user action (unpausing a document).

Here, because we're setting `docConfig.isPaused = false` via a setter, we can't throw, so if the underlying update can't be completed, we're passing the error through the error listener, so the handling of this error is going to happen somewhere else entirely on another thread/dispatch queue. This seems kind of unexpected, especially considering the error occurred as a direct result of something that the user directly initiated.

At the same time, isPaused can be set on a sync pass, which is not user initiated, so in those cases the FatalErrorListener might make sense. Right now on Android, those errors get caught at the `DataSynchronizerRunner.run()` method where they get logged and squashed. 

I'm starting to think that what we really might want is for almost of our underlying methods to throw rather than notify a fatal error listener, and only notify on the fatal error listener when the error is caught in a non-user initiated place like the DataSynchronizerRunner. That way, user-initiated actions that fail will throw immediately, and other actions will notify.

Reply:
I feel that consistency is more important -- that is, I don't care if we throw all or capture all with a listener, but we need to be consistent within the constraints of our dichotomy...

On that point I also would argue the dichotomy ought not be user vs internal, but recoverable vs non-recoverable

Reply:
If we want to take this to its furthest point, then irrecoverable errors need to stop the DataSynchronizer entirely. Which I think is the way to go. We should make a ticket for this.

Reply:
As a note, on Android, these "fatal errors" are actually swallowed. This is currently unique to iOS.